### ISSUE TITLE
hierarchical binding for vcml::tlm_target_socket: "(E126) sc_export instance already bound"

### ISSUE THREAD

**Issue by soundart** [machineware-gmbh/vcml#65]
Hi,

I have a question wrt to hierarchical binding of vcml::tlm_target_socket. 

The situation I have is:
I have a child module ("stream") with a vcml::tlm_target_socket ("stream_child_socket") and a parent module ("dma") with a vcml::tlm_target_socket.

If I bind now the parent-target-socket to the child-target-socket

`  
      parent_socket.bind(stream.stream_child_socket);
`

  I get the following error during elaboraton:

```
C++ exception with description "Error: (E126) sc_export instance already bound: tb.dma.stream.stream_child_socket
In file: /build/3rdparty/systemc/install/include/sysc/communication/sc_export.h:188" thrown in the test body.
```

What is the best approach to deal with this? 

The SystemC language reference seems to indicate this behaviour is intentional for simple_target_sockets in table "Table 60—Permitted socket bindings". If I read this correctly.

But then I am wondering what the purpose of this template specialisation is:

https://github.com/machineware-gmbh/vcml/blame/8b21479538a35c7549dfe278e6dc71465619e46e/include/vcml/protocols/tlm_sockets.h#L372

**Comment by janweinstock** [comment]
To do a hierarchical TLM binding between the `dma` parent module and `stream` submodule, you need to use `tlm::tlm_target_socket<>` in the parent module (instead of a `vcml::tlm_target_socket`). 

Then simply bind your `stream.stream_child_socket` to this new target socket.

**Comment by soundart** [comment]
Ok, but then I loose the vcml convenience functions like stub() and maybe trace()?

**Comment by janweinstock** [comment]
Correct. But you can still manually stub the socket using a `vcml::tlm_initiator_stub`.

**Comment by shesselba** [comment]
How about having another specialization that adds an adapter if vcml::tlm_target_socket is connected to another vcml::tlm_target_socket even if they have the same width?

**Comment by janweinstock** [comment]
I think the correct way to handle this would be to move the desired functionality (stub, trace, ...) into a `tlm_base_target_socket`. We already do it like this for the other protocols (gpio, clk, ...) to allow stubbing of hierarchically bounds ports. In this context, "base" sockets support stubbing and adapters, but they can only forward transactions, not receive them.

We did not use this approach for normal TLM sockets, because we felt having the convenience sockets (e.g. `simple_target_socket`) would be more useful than building our own base sockets and then derive from them.

Having an adapter would surely work, but it introduces a lot of overhead (one extra `sc_module`) that could be avoided by using a hierarchical binding like it is intended by the standard.

**Comment by aut0** [comment]
Tracing should still work for the child socket.
